File: backend/src/modules/system/system.service.ts
import { Injectable } from '@nestjs/common';
import { PrismaService } from '@/infrastructure/prisma/prisma.service';

@Injectable()
export class SystemService {
  constructor(private prisma: PrismaService) {}

  async getRoles() {
    return this.prisma.role.findMany({
      include: {
        permissions: { include: { permission: true } },
        _count: { select: { users: true } },
      },
    });
  }

  async getPermissions() {
    return this.prisma.permission.findMany({ orderBy: { module: 'asc' } });
  }

  async getAuditLogs(params: { page: number; pageSize: number; module?: string }) {
    const { page, pageSize, module } = params;
    const where = module ? { module } : {};

    const [data, total] = await Promise.all([
      this.prisma.auditLog.findMany({
        where,
        skip: (page - 1) * pageSize,
        take: pageSize,
        include: { user: { select: { id: true, name: true, email: true } } },
        orderBy: { createdAt: 'desc' },
      }),
      this.prisma.auditLog.count({ where }),
    ]);

    return { data, total, page, pageSize, totalPages: Math.ceil(total / pageSize) };
  }

  async getEmailTemplates() {
    return this.prisma.emailTemplate.findMany({ orderBy: { name: 'asc' } });
  }

  async updateEmailTemplate(id: string, data: { subject?: string; body?: string; isActive?: boolean }) {
    return this.prisma.emailTemplate.update({ where: { id }, data });
  }
}

File: backend/src/main/java/com/company/hr/service/SystemService.java
package com.company.hr.service;

import com.company.hr.common.dto.PagedResponse;
import com.company.hr.entity.AuditLog;
import com.company.hr.entity.EmailTemplate;
import com.company.hr.entity.Permission;
import com.company.hr.entity.Role;
import com.company.hr.repository.AuditLogRepository;
import com.company.hr.repository.EmailTemplateRepository;
import com.company.hr.repository.PermissionRepository;
import com.company.hr.repository.RoleRepository;
import jakarta.persistence.EntityNotFoundException;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@RequiredArgsConstructor
public class SystemService {

    private final RoleRepository roleRepository;
    private final PermissionRepository permissionRepository;
    private final AuditLogRepository auditLogRepository;
    private final EmailTemplateRepository emailTemplateRepository;

    @Transactional(readOnly = true)
    public List<Role> getRoles() {
        return roleRepository.findAll();
    }

    @Transactional(readOnly = true)
    public List<Permission> getPermissions() {
        return permissionRepository.findAll();
    }

    @Transactional(readOnly = true)
    public PagedResponse<AuditLog> getAuditLogs(int page, int pageSize, String module) {
        Pageable pageable = PageRequest.of(page - 1, pageSize, Sort.by("createdAt").descending());
        Page<AuditLog> result;
        if (module != null && !module.isBlank()) {
            result = auditLogRepository.findByModule(module, pageable);
        } else {
            result = auditLogRepository.findAll(pageable);
        }
        return PagedResponse.from(result);
    }

    @Transactional(readOnly = true)
    public List<EmailTemplate> getEmailTemplates() {
        return emailTemplateRepository.findAll();
    }

    @Transactional
    public EmailTemplate updateEmailTemplate(String id, String subject, String body, Boolean isActive) {
        EmailTemplate template = emailTemplateRepository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("Email template not found: " + id));
        if (subject != null) {
            template.setSubject(subject);
        }
        if (body != null) {
            template.setBody(body);
        }
        if (isActive != null) {
            template.setActive(isActive);
        }
        return emailTemplateRepository.save(template);
    }
}
